src/node/txorphanage_impl.h

```diff
@@ -0,0 +1,672 @@
+     * amount of announcements and space for each peer. The reserved amount is protected from eviction even if there
+     * are peers spamming the orphanage.
+     */
+    void LimitOrphans()
+    {
+        if (m_orphans.empty() || !NeedsTrim()) return;
+
+        const auto original_unique_txns{CountUniqueOrphans()};
+
+        // These numbers cannot change within a single call to LimitOrphans because the size of m_peer_orphanage_info
+        // does not change unless a peer is removed.
+        const auto max_ann{MaxPeerAnnouncements()};
+        const auto max_mem{ReservedPeerUsage()};
+
+        // We have exceeded the global limit(s). Now, identify who is using too much and evict their orphans.
+        // Create a heap of pairs (NodeId, DoS score), sorted by descending DoS score.
+        std::vector<std::pair<NodeId, FeeFrac>> heap_peer_dos;
+        heap_peer_dos.reserve(m_peer_orphanage_info.size());
+        for (const auto& [nodeid, entry] : m_peer_orphanage_info) {
+            heap_peer_dos.emplace_back(nodeid, entry.GetDosScore(max_ann, max_mem));
+        }
+        auto compare_score = [](auto left, auto right) { return left.second < right.second; };
+        std::make_heap(heap_peer_dos.begin(), heap_peer_dos.end(), compare_score);
+
+        unsigned int num_erased{0};
+        // This loop runs a maximum number of MaxGlobalAnnouncements() iterations.
+        do {
+            Assume(!heap_peer_dos.empty());
+            // Find the peer with the highest DoS score, which is a fraction of {usage, announcements} used over the
+            // respective allowances. This metric causes us to naturally select peers who have exceeded their limits
+            // (i.e. a DoS score > 1) before peers who haven't, and the loop should halt before we ever select peers who
+            // haven't. We may choose the same peer as the last iteration of this loop.
+            // Note: if ratios are the same, FeeFrac tiebreaks by denominator. In practice, since the CPU denominator is
+            // always lower, this means that a peer with only high number of announcements will be targeted before a
+            // peer using a lot of memory, even if they have the same ratios.
+            std::pop_heap(heap_peer_dos.begin(), heap_peer_dos.end(), compare_score);
+            const auto [worst_peer, dos_score] = std::move(heap_peer_dos.back());
+            heap_peer_dos.pop_back();
+
+            // If needs trim, then at least one peer has a DoS score higher than 1.
+            Assume(dos_score.fee > dos_score.size);
+
+            // Evict the oldest announcement from this peer, sorting non-reconsiderable before reconsiderable.
+            auto it_ann = m_orphans.get<ByPeer>().lower_bound(ByPeerView{worst_peer, false, 0});
+            Assume(it_ann->m_announcer == worst_peer);
+            Erase<ByPeer>(it_ann, /*cleanup_outpoints_map=*/CountWtxid(it_ann->m_tx->GetWitnessHash()) == 1);
```

Comment:
If you had `MaxGlobalAnnouncements` peers all announcing the same tx, it would end up being `n^2 * logn` work with `CountWtxid` calls?

Reply:
Oof yes! Should be replaced with an "IsUnique" type of thing.